File: 61.php
<?php

/**
 * Definition for a singly-linked list.
 * class ListNode {
 *     public $val = 0;
 *     public $next = null;
 *     function __construct($val) { $this->val = $val; }
 * }
 */
class Solution {

  /**
   * @param ListNode $head
   * @param Integer $k
   * @return ListNode
   */
  function rotateRight($head, $k) {
    if (!$head || $k == 0) {
      return $head;
    }
    $length = 1;
    $cur = $head;
    while ($cur->next) {
      $cur = $cur->next;
      $length++;
    }

    $cur->next = $head;
    $tail = $cur;
    $cur = $head;

    $cycleTime = $length - ($k % $length);
    for ($i = 0; $i < $cycleTime; $i++) {
      $tail = $tail->next;
      $cur = $cur->next;
    }
    $tail->next = null;
    return $cur;
  }
}

File: go/linkedList/61.go
package main
//
//func main() {
//	n1 := &ListNode{Val: 1}
//	n2 := &ListNode{Val: 2}
//	n3 := &ListNode{Val: 3}
//	n4 := &ListNode{Val: 4}
//	n5 := &ListNode{Val: 5}
//	n1.Next = n2
//	n2.Next = n3
//	n3.Next = n4
//	n4.Next = n5
//
//	rotateRight(n1, 2)
//}
//
//type ListNode struct {
//	Val  int
//	Next *ListNode
//}
//
//func rotateRight(head *ListNode, k int) *ListNode {
//	if head == nil || head.Next == nil {
//		return head
//	}
//
//	length, tail := 1, head
//	for tail.Next != nil {
//		tail = tail.Next
//		length++
//	}
//	tail.Next = head
//	cycleTime := length - (k % length)
//	for cycleTime > 0 {
//		cycleTime--
//		head, tail = head.Next, tail.Next
//	}
//	tail.Next = nil
//	return head
//}
